### PR TITLE
SEQNG-887: Make Gnirs epics reads referentially transparent

### PR DESCRIPTION
**Pull request by cquiroz**
Another PR making the Epics reads referentially transparent. This time is GNIRS

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala [hunk 1]
```diff
@@ -238,73 +241,84 @@ object GnirsControllerEpics {
...
 
-  def apply(): GnirsController[IO] =
+  val checkDhs =
+    failUnlessM(
+      epicsSys.dhsConnected.map(_.forall(identity)),
+                SeqexecFailure.Execution("GNIRS is not connected to DHS"))
+
+  val checkArray =
+    failUnlessM(
+      epicsSys.arrayActive.map(_.forall(identity)),
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: @jluhrs Previously `arrayActive` was checked with `exists`. Was there a reason for that?

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala [hunk 2]
```diff
@@ -126,100 +133,99 @@ class GnirsEpics(epicsService: CaService, tops: Map[String, String]) {
...
 
-  val state: CaStatusAcceptor = epicsService.getStatusAcceptor("nirs::status")
-  val dcState: CaStatusAcceptor = epicsService.getStatusAcceptor("nirs::dcstatus")
-
-  def arrayId: Option[String] = Option(dcState.getStringAttribute("arrayid").value)
+  private val state: CaStatusAcceptor = epicsService.getStatusAcceptor("nirs::status")
+  private val dcState: CaStatusAcceptor = epicsService.getStatusAcceptor("nirs::dcstatus")
 
-  def arrayType: Option[String] = Option(dcState.getStringAttribute("arraytyp").value)
+  def arrayId: F[Option[String]] = safeAttribute(dcState.getStringAttribute("arrayid"))
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala]: We are leaning to produce items with the form `F[A]` rather than `F[Option[A]]` but I want to do than in a later PR

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala [hunk 3]
```diff
@@ -126,100 +133,99 @@ class GnirsEpics(epicsService: CaService, tops: Map[String, String]) {
...
+    .map(_.toInt =!= 0).value
 
   val arrayActiveAttr: Option[CaAttribute[JDetectorState]] = Option(dcState.addEnum(
     "arrayState", s"${GnirsTop}dc:activate", classOf[JDetectorState]
   ))
-  def arrayActive: Option[Boolean] = arrayActiveAttr.flatMap(at => Option(at.value))
-    .map(_.getActive)
 
-  def minInt: Option[Double] = Option(dcState.getDoubleAttribute("minInt").value).map(_.toDouble)
+  def arrayActive: F[Option[Boolean]] =
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala]: @jluhrs Could you review this one in particular. I'm not sure if I fully understand the original

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala [hunk 4]
```diff
@@ -103,17 +92,25 @@ object GnirsControllerEpics {
...
       case Acquisition => "In"
       case _           => "Out"
     }
 
-    smartSetParam(v, epicsSys.acqMirror.map(removePartName), ccCmd.setAcqMirror(v))
+    smartSetParamF(v, Nested(epicsSys.acqMirror).map(removePartName).value, ccCmd.setAcqMirror(v))
   }
 
-  private def setGrating(s: Spectrography, c: Camera): List[SeqAction[Unit]] = {
+  private def setGrating(s: Spectrography, c: Camera): List[IO[Option[IO[Unit]]]] = {
```
Comment (jluhrs) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: In these case an the ones below, couldn't `IO[List[IO[Unit]]]` be used instead of `List[IO[Option[IO[Unit]]]]`.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala (hunk 3 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsEpics.scala]: The original takes the attribute, which reads a binary enum, and converts it to a `Boolean`. Of course, the attribute could not exists, that is why it is treated a an Option. Your implementation follows that, but I don't understand the last `_.getOrElse(false)`. There is no default value in the original. If the value cannot be read, then the whole result is `None`. That us why there is a `flatMap`.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala (hunk 1 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: It was `GnirsEpics.instance.arrayActive.exists(identity)`, meaning that to pass, the value must be read properly and must be true. (i.e. it must be `Some(true)`). Now `arrayActive` gives a `F[Option[Boolean]]`. `forall` is true for `None`, therefore your code will not stop the observation if it cannot verify the state of the array. That is wrong.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala (hunk 4 above)
Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: I've tried to implement this on a different way but the issue I found is that I need to know if there are elements on the list. If there are no elements on the list I shouldn't execute the `after` part on https://github.com/gemini-hlsw/ocs3/blob/develop/modules/seqexec/server/src/main/scala/seqexec/server/EpicsUtil.scala#L426
That is needed to avoid calling `post` if there are no changes.
I can still look for a better way to do this. I agree it is hard to follow

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala (hunk 1 above)
Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: It seems we should we apply the same logic for the DHS checks, WDYT?

Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: Yes. GNIRS can crash if you try to take an image when it is not connected to the DHS or the detector array is not active.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala [hunk 5]
```diff
@@ -238,73 +241,84 @@ object GnirsControllerEpics {
...
+                SeqexecFailure.Execution("GNIRS is not connected to DHS"))
+
+  val checkArray =
+    failUnlessM(
+      epicsSys.arrayActive.map(_.contains(true)),
+                SeqexecFailure.Execution("GNIRS detector array is not active"))
+
+  val warnOnDhs =
+    epicsSys.dhsConnected.map(_.contains(true)).flatMap(
+       IO(Log.warn("GNIRS is not connected to DHS")).whenA
```
Comment (jluhrs) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: `.unlessA` ?

Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: You're right
